src/optimizer/index.js

```diff
@@ -49,9 +52,18 @@ module.exports = {
 
         const transformer = optimizationTransforms[transformName];
 
-        result = transform.transform(regexp, transformer);
-        regexp = result.getAST();
+        let newResult = transform.transform(ast, transformer);
+
+        if (newResult.toString() !== result.toString()) {
+          if (newResult.toString().length <= result.toString().length) {
+            result = newResult;
+          } else {
+            // Result has changed but is not shorter: restore ast to its previous state.
+            ast = clone(result.getAST());
```

Comment:
So it seems to me, we can just re-parse in this case (since we already were doing this in the main `do-while` loop, previous line 38, where `regexp = prevResult;`, and which was the string).

The `clone` is a very generic clone function, which considers several edges cases (like cyclic references, different types of objects, etc). Plus, we're introducing another dependency.

I think we can try just doing here the:

```js
regexp = result.toString();
```

and it can normally be fed further to the `transform.transform(...)`. Re-parsing is specific to this structure, and potentially might be faster than the generic clone on larger code-bases.

I haven't tested it yet, but how about this:

```js
optimize(regexp, transformsWhitelist = []) {
    const transformToApply = transformsWhitelist.length > 0
      ? transformsWhitelist
      : Object.keys(optimizationTransforms);

    let prevResult = new transform.TransformResult(
      typeof regexp === 'object' ? regexp
      : parser.parse(`${regexp}`);
    );

    let result;

    do {
      transformToApply.forEach(transformName => {

        if (!optimizationTransforms.hasOwnProperty(transformName)) {
          throw new Error(
            `Unknown optimization-transform: ${transformName}. ` +
            `Available transforms are: ` +
            Object.keys(optimizationTransforms).join(', ')
          );
        }

        const transformer = optimizationTransforms[transformName];

        // Obtain the next result.
        result = transform.transform(regexp, transformer);

        // New transformation resulted to a longer regexp, restore:
        if (prevResult.toString().length < result.toString().length) {
          regexp = prevResult.toString();
        } else {
          // Otherwise, proceed with the same AST:
          regexp = result.getAST();
        }

        prevResult = result;
      });
    } while (result.toString() !== prevResult.toString);

    return result;
  },
```

Reply:
@DmitrySoshnikov Ok, so I gave it a try, basically replacing the call to `clone()` with `result.toString()`. This does not help at all with the performance :cry: 
All my previous tests take 2 to 3 times longer to run!

I think parsing the regexp takes much more time than cloning the AST.

Regarding the new dependency and the unneeded complexity, would you be ok with a custom and simple deep copy implementation?

Reply:
Hm.. interesting. Yeah, if the `clone` implemented correctly, it might be better, and in this case we can introduce this dependency I think. OK, sounds good, I'll take another look at the evening, and will merge in case. Thanks for the tests!

Reply:
I just noticed the clone library I picked in this PR has a second parameter than can be set to `false` to disable the handling of circular data to improve performance.

> This function supports circular references by default, but if you are certain
> there are no circular references in your object, you can save some CPU time
> by calling clone(obj, false).
> Caution: if `circular` is false and `parent` contains circular references,
> your program may enter an infinite loop and crash.

Maybe we can take advantage of this, since the AST has no reason to have circular references, right?

Reply:
Yeah, also, that's right. Also, the only JS objects which can appear in AST are `{}`, and `[]`, i.e. not dates, functions, etc. So from this perspective, a custom and simple clone utility as you mentioned might a good alternative actually, since we will not need to handle all those edge case.

Reply:
If we go with the custom clone util, we can put it in `src/utils/clone.js`. And we can just take basic functionality from the `clone` package, and remove all edges cases making it a simple for the use-case.